### PR TITLE
[MF-284] Align 'Edit' button to work correctly with esm-config's navigator used in the patient registration component

### PR DESCRIPTION
**Pull request by samuelmale**
**Parent PR**
> https://github.com/openmrs/openmrs-esm-patient-registration/pull/51

**Ticket**
> https://issues.openmrs.org/browse/MF-284

### CODE REVIEW

**Comment by brandones** [comment]
EDIT: whoops, just saw that this is a different repo from https://github.com/openmrs/openmrs-esm-patient-registration/pull/51, sorry! Lemme look.

**Comment by brandones** [comment]
I think it should be the responsibility of whoever *sets* the `afterUrl` to include the correct URL base. Does that make sense?

**Comment by brandones** [comment]
Wait, sorry, it's taken me a minute to figure out what's going on here. Here, you are in fact setting `afterUrl`, so my previous comment doesn't make sense. However, we should never be passing magic template stuff in URL parameters like this. Does the existing code not work correctly? It seems like this change shouldn't be necessary at all.

**Comment by samuelmale** [comment]
@brandones The existing code works correctly as is. But as soon as we have https://github.com/openmrs/openmrs-esm-patient-registration/pull/51 merged in, it will break. This is because the `afterUrl` won't contain the `openmrsSpaBase` yet the [navigate({ to: string })](https://github.com/openmrs/openmrs-esm-module-config/blob/2c014c2f954ad700f9be0eec1f1ccc7d8d8d5aee/src/navigation/navigate.ts#L24-L33) expects the URL to have the base or rather something close to an absolute path hence the redirect will be broken. 

In short, in the parent PR I replaced `history.push(..)` from react-router with `navigate(..)` from esm-config.

**Comment by brandones** [comment]
`navigate` can accept any kind of path. Passing it a root-relative path will behave as expected. `afterUrl` should not accept template parameters.

EDIT: Or rather, we should treat `afterUrl` as if it does not accept template parameters.

File: src/widgets/banner/patient-banner.component.tsx [hunk 1]
```diff
@@ -67,7 +67,7 @@ export default function PatientBanner(props: PatientBannerProps) {
               to={
                 "/patient-registration/patient/" +
                 patient.id +
-                "?afterUrl=" +
+                "?afterUrl=${openmrsSpaBase}" +
```
Comment (brandones) [inline src/widgets/banner/patient-banner.component.tsx]:
```suggestion
                "?afterUrl=" +
                window.getOpenmrsSpaBase() +
```

**Comment by samuelmale** [comment]
@brandones applied commit suggestion. Thanks!

**Comment by brandones** [comment]
No problem. Looks like there's a Typescript failure. Try `(window as any).getOpenmrsSpaBase()`

**Comment by samuelmale** [comment]
> No problem. Looks like there's a Typescript failure. Try `(window as any).getOpenmrsSpaBase()`

Oya, but I have some early findings:

- `window.getOpenmrsSpaBase()` returns `/openmrs/spa/`
- `pathname` from the `useLocation()` hook will be `/patient/54c77c2c-a53f-41ca-a167-705ae0141b9f/chart/summary`

Appending the base to the pathname results to something like: `/openmrs/spa//patient/54c77c2c-a53f-41ca-a167-705ae0141b9f/chart/summary`

There are trailing slashes: `spa//patient`

**Comment by samuelmale** [comment]
@brandones I think this is why this exists: https://github.com/openmrs/openmrs-esm-module-config/blob/2c014c2f954ad700f9be0eec1f1ccc7d8d8d5aee/src/navigation/navigate.ts#L25

**Comment by samuelmale** [comment]
@brandones Fixed.

**Review by brandones — APPROVED** [review]
Awesome, thanks Samuel! Yeah, getting the current root path could be a little easier than it presently is. Maybe something we'll add to esm-core at some point.
